File: design-to-spec/scripts/validate-contracts.py
#!/usr/bin/env python3
"""Validate design-to-spec YAML contracts.

This checks the cross-file references that commonly drift between UI_Schema,
API_Schema, and Mapping_Logic before the final spec assembly step.
"""

from __future__ import annotations

import argparse
import sys
from pathlib import Path
from typing import Any

try:
    import yaml
except ImportError:
    print("PyYAML is required: python3 -m pip install pyyaml", file=sys.stderr)
    sys.exit(2)


def load_yaml(path: Path) -> dict[str, Any]:
    try:
        data = yaml.safe_load(path.read_text(encoding="utf-8"))
    except yaml.YAMLError as exc:
        raise ValueError(f"{path}: invalid YAML: {exc}") from exc
    if not isinstance(data, dict):
        raise ValueError(f"{path}: top-level YAML must be a mapping")
    return data


def collect_api_ids(api: dict[str, Any]) -> tuple[set[str], set[str], set[str]]:
    endpoints = api.get("api", {}).get("endpoints", []) or []
    endpoint_ids: set[str] = set()
    params: set[str] = set()
    fields: set[str] = set()
    for endpoint in endpoints:
        endpoint_id = endpoint.get("id")
        if endpoint_id:
            endpoint_ids.add(endpoint_id)
        for param in endpoint.get("params", []) or []:
            name = param.get("name")
            if name:
                params.add(name)
        for field in endpoint.get("response_fields", []) or []:
            name = field.get("name")
            if name:
                fields.add(name)
    return endpoint_ids, params, fields


def validate(ui_doc: dict[str, Any], api_doc: dict[str, Any], mapping_doc: dict[str, Any]) -> list[str]:
    errors: list[str] = []
    ui = ui_doc.get("ui", {})
    api = api_doc.get("api", {})
    mapping = mapping_doc.get("mapping", {})

    if not isinstance(ui, dict):
        errors.append("ui-schema: missing ui mapping")
        ui = {}
    if not isinstance(api, dict):
        errors.append("api-schema: missing api mapping")
        api = {}
    if not isinstance(mapping, dict):
        errors.append("mapping-logic: missing mapping mapping")
        mapping = {}

    component_name = ui.get("name")
    if component_name and mapping.get("component") and mapping.get("component") != component_name:
        errors.append(f"mapping.component {mapping.get('component')!r} does not match ui.name {component_name!r}")

    components = ui.get("components", []) or []
    component_ids = {item.get("id") for item in components if item.get("id")}
    state_ids = {item.get("id") for item in ui.get("states", []) or [] if item.get("id")}
    state_assertions = {
        item.get("id"): item.get("render_assertion")
        for item in ui.get("states", []) or []
        if item.get("id")
    }

    for item in components:
        parent_id = item.get("parent_id")
        if parent_id and parent_id != "root" and parent_id not in component_ids:
            errors.append(f"ui.components[{item.get('id')}].parent_id references missing component {parent_id!r}")

    endpoint_ids, param_names, response_fields = collect_api_ids(api_doc)
    requests = mapping.get("data_fetching", {}).get("requests", []) or []
    request_ids = {item.get("id") for item in requests if item.get("id")}
    for request in requests:
        endpoint = request.get("endpoint")
        if endpoint and endpoint_ids and endpoint not in endpoint_ids:
            errors.append(f"request {request.get('id')!r} references missing endpoint {endpoint!r}")
        for dependency in request.get("depends_on", []) or []:
            if dependency not in request_ids:
                errors.append(f"request {request.get('id')!r} depends on missing request {dependency!r}")

    for binding in mapping.get("bindings", []) or []:
        direction = binding.get("direction")
        if direction == "ui_to_api":
            source_ui = binding.get("source_ui")
            target_api = binding.get("target_api")
            if source_ui not in component_ids:
                errors.append(f"binding source_ui references missing component {source_ui!r}")
            if target_api not in param_names:
                errors.append(f"binding target_api references missing API param {target_api!r}")
        elif direction == "api_to_ui":
            source_api = binding.get("source_api")
            target_ui = binding.get("target_ui")
            if source_api not in response_fields:
                errors.append(f"binding source_api references missing response field {source_api!r}")
            if target_ui not in component_ids:
                errors.append(f"binding target_ui references missing component {target_ui!r}")
        elif direction == "ui_to_event":
            source_ui = binding.get("source_ui")
            target_event = binding.get("target_event")
            if source_ui not in component_ids:
                errors.append(f"binding source_ui references missing component {source_ui!r}")
            if not target_event:
                errors.append("ui_to_event binding is missing target_event")
        else:
            errors.append(f"binding has unsupported direction {direction!r}")

    for transition in mapping.get("state_machine", []) or []:
        to_state = transition.get("to")
        if to_state not in state_ids:
            errors.append(f"state_machine transition points to missing state {to_state!r}")
        if not transition.get("render_assertion") and not state_assertions.get(to_state):
            errors.append(f"state_machine transition to {to_state!r} has no render_assertion fallback")

    for state in ui.get("states", []) or []:
        if state.get("required") is True and not state.get("render_assertion"):
            errors.append(f"required state {state.get('id')!r} is missing render_assertion")

    return errors


def main() -> int:
    parser = argparse.ArgumentParser(description="Validate design-to-spec YAML contracts.")
    parser.add_argument("--ui", required=True, type=Path, help="Path to ui-schema.yaml")
    parser.add_argument("--api", required=True, type=Path, help="Path to api-schema.yaml")
    parser.add_argument("--mapping", required=True, type=Path, help="Path to mapping-logic.yaml")
    args = parser.parse_args()

    try:
        ui_doc = load_yaml(args.ui)
        api_doc = load_yaml(args.api)
        mapping_doc = load_yaml(args.mapping)
        errors = validate(ui_doc, api_doc, mapping_doc)
    except ValueError as exc:
        print(str(exc), file=sys.stderr)
        return 1

    if errors:
        for error in errors:
            print(f"ERROR: {error}", file=sys.stderr)
        return 1

    print("OK: contracts are valid")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: skills/design-to-spec/scripts/validate-contracts.js
#!/usr/bin/env node
// Validate design-to-spec YAML contracts: schema, uniqueness, and cross-refs.

import { readFileSync } from 'node:fs';
import { dirname, resolve } from 'node:path';
import { fileURLToPath } from 'node:url';
import { loadYaml } from './lib/yaml.js';
import { validateJsonSchema } from './lib/json-schema.js';
import { parseArgs, requireOpts } from './lib/cli.js';

const __dirname = dirname(fileURLToPath(import.meta.url));
const SCHEMA_DIR = resolve(__dirname, '..', 'schemas');

function repr(value) {
  if (typeof value === 'string') return `'${value.replace(/'/g, "\\'")}'`;
  if (value === null || value === undefined) return 'None';
  return String(value);
}

function loadJson(path) {
  let text;
  try {
    text = readFileSync(path, 'utf8');
  } catch (err) {
    throw new Error(`${path}: cannot read: ${err.message}`);
  }
  let data;
  try {
    data = JSON.parse(text);
  } catch (err) {
    throw new Error(`${path}: invalid JSON: ${err.message}`);
  }
  if (data === null || typeof data !== 'object' || Array.isArray(data)) {
    throw new Error(`${path}: top-level JSON must be a mapping`);
  }
  return data;
}

function validateSchema(label, document, schemaPath) {
  return validateJsonSchema(label, document, loadJson(schemaPath));
}

function findDuplicates(items) {
  const seen = new Set();
  const duplicates = new Set();
  for (const item of items) {
    if (seen.has(item)) duplicates.add(item);
    else seen.add(item);
  }
  return duplicates;
}

function validateUniqueValues(label, values, errors) {
  const filtered = values.filter((v) => v);
  const dups = [...findDuplicates(filtered)].sort();
  for (const dup of dups) {
    errors.push(`${label} contains duplicate value ${repr(dup)}`);
  }
}

function collectApiIds(apiDoc) {
  const endpoints = apiDoc.api?.endpoints ?? [];
  const endpointIds = new Set();
  const params = new Set();
  const fields = new Set();
  for (const endpoint of endpoints) {
    if (endpoint.id) endpointIds.add(endpoint.id);
    for (const p of endpoint.params ?? []) {
      if (p.name) params.add(p.name);
    }
    for (const b of endpoint.request_body ?? []) {
      if (b.name) params.add(b.name);
    }
    for (const f of endpoint.response_fields ?? []) {
      if (f.name) fields.add(f.name);
    }
  }
  return { endpointIds, params, fields };
}

function validateUniqueness(ui, api, mapping) {
  const errors = [];
  const components = ui.components ?? [];
  const states = ui.states ?? [];
  validateUniqueValues(
    'ui.components[].id',
    components.map((i) => i.id),
    errors,
  );
  validateUniqueValues(
    'ui.states[].id',
    states.map((i) => i.id),
    errors,
  );

  const endpoints = api.endpoints ?? [];
  validateUniqueValues(
    'api.endpoints[].id',
    endpoints.map((i) => i.id),
    errors,
  );
  for (const endpoint of endpoints) {
    const eid = endpoint.id ?? '<unknown>';
    const label = `api.endpoints[${eid}]`;
    validateUniqueValues(
      `${label}.params[].name`,
      (endpoint.params ?? []).map((i) => i.name),
      errors,
    );
    validateUniqueValues(
      `${label}.response_fields[].name`,
      (endpoint.response_fields ?? []).map((i) => i.name),
      errors,
    );
  }

  validateUniqueValues(
    'api.open_questions[].id',
    (api.open_questions ?? []).map((i) => i.id),
    errors,
  );

  const requests = mapping.data_fetching?.requests ?? [];
  validateUniqueValues(
    'mapping.data_fetching.requests[].id',
    requests.map((i) => i.id),
    errors,
  );
  validateUniqueValues(
    'mapping.open_questions[].id',
    (mapping.open_questions ?? []).map((i) => i.id),
    errors,
  );

  return errors;
}

function validate(uiDoc, apiDoc, mappingDoc) {
  const errors = [];
  let ui = uiDoc.ui ?? {};
  let api = apiDoc.api ?? {};
  let mapping = mappingDoc.mapping ?? {};

  if (typeof ui !== 'object' || ui === null || Array.isArray(ui)) {
    errors.push('ui-schema: missing ui mapping');
    ui = {};
  }
  if (typeof api !== 'object' || api === null || Array.isArray(api)) {
    errors.push('api-schema: missing api mapping');
    api = {};
  }
  if (typeof mapping !== 'object' || mapping === null || Array.isArray(mapping)) {
    errors.push('mapping-logic: missing mapping mapping');
    mapping = {};
  }

  const componentName = ui.name;
  if (componentName && mapping.component && mapping.component !== componentName) {
    errors.push(
      `mapping.component ${repr(mapping.component)} does not match ui.name ${repr(componentName)}`,
    );
  }

  errors.push(...validateUniqueness(ui, api, mapping));

  const components = ui.components ?? [];
  const componentIds = new Set(components.map((i) => i.id).filter(Boolean));
  const states = ui.states ?? [];
  const stateIds = new Set(states.map((i) => i.id).filter(Boolean));
  const stateAssertions = new Map();
  for (const state of states) {
    if (state.id) stateAssertions.set(state.id, state.render_assertion);
  }

  for (const item of components) {
    const parentId = item.parent_id;
    if (parentId && parentId !== 'root' && !componentIds.has(parentId)) {
      errors.push(
        `ui.components[${item.id}].parent_id references missing component ${repr(parentId)}`,
      );
    }
  }

  for (const state of states) {
    for (const scoped of state.scope_components ?? []) {
      if (!componentIds.has(scoped)) {
        errors.push(
          `ui.states[${state.id}].scope_components references missing component ${repr(scoped)}`,
        );
      }
    }
  }

  const { endpointIds, params: paramNames, fields: responseFields } = collectApiIds(apiDoc);
  const requests = mapping.data_fetching?.requests ?? [];
  const requestIds = new Set(requests.map((i) => i.id).filter(Boolean));
  for (const request of requests) {
    const endpoint = request.endpoint;
    if (endpoint && endpointIds.size > 0 && !endpointIds.has(endpoint)) {
      errors.push(`request ${repr(request.id)} references missing endpoint ${repr(endpoint)}`);
    }
    for (const dep of request.depends_on ?? []) {
      if (!requestIds.has(dep)) {
        errors.push(`request ${repr(request.id)} depends on missing request ${repr(dep)}`);
      }
    }
  }

  for (const binding of mapping.bindings ?? []) {
    const direction = binding.direction;
    if (direction === 'ui_to_api') {
      if (!componentIds.has(binding.source_ui)) {
        errors.push(`binding source_ui references missing component ${repr(binding.source_ui)}`);
      }
      if (!paramNames.has(binding.target_api)) {
        errors.push(`binding target_api references missing API param ${repr(binding.target_api)}`);
      }
    } else if (direction === 'api_to_ui') {
      if (!responseFields.has(binding.source_api)) {
        errors.push(
          `binding source_api references missing response field ${repr(binding.source_api)}`,
        );
      }
      if (!componentIds.has(binding.target_ui)) {
        errors.push(`binding target_ui references missing component ${repr(binding.target_ui)}`);
      }
    } else if (direction === 'ui_to_event') {
      if (!componentIds.has(binding.source_ui)) {
        errors.push(`binding source_ui references missing component ${repr(binding.source_ui)}`);
      }
      if (!binding.target_event) {
        errors.push('ui_to_event binding is missing target_event');
      }
    } else {
      errors.push(`binding has unsupported direction ${repr(direction)}`);
    }
  }

  for (const transition of mapping.state_machine ?? []) {
    const toState = transition.to;
    if (!stateIds.has(toState)) {
      errors.push(`state_machine transition points to missing state ${repr(toState)}`);
    }
    if (!transition.render_assertion && !stateAssertions.get(toState)) {
      errors.push(`state_machine transition to ${repr(toState)} has no render_assertion fallback`);
    }
  }

  for (const state of states) {
    if (state.required === true && !state.render_assertion) {
      errors.push(`required state ${repr(state.id)} is missing render_assertion`);
    }
  }

  return errors;
}

function main() {
  let args;
  try {
    args = parseArgs(process.argv.slice(2), {
      options: ['ui', 'api', 'mapping', 'ui-schema', 'api-schema', 'mapping-schema'],
    });
    requireOpts(args, ['ui', 'api', 'mapping']);
  } catch (err) {
    console.error(err.message);
    return 1;
  }

  const uiSchemaPath = args['ui-schema'] ?? resolve(SCHEMA_DIR, 'ui-schema.json');
  const apiSchemaPath = args['api-schema'] ?? resolve(SCHEMA_DIR, 'api-schema.json');
  const mappingSchemaPath = args['mapping-schema'] ?? resolve(SCHEMA_DIR, 'mapping-logic.json');

  let errors = [];
  try {
    const uiDoc = loadYaml(args.ui);
    const apiDoc = loadYaml(args.api);
    const mappingDoc = loadYaml(args.mapping);
    errors.push(...validateSchema('ui-schema', uiDoc, uiSchemaPath));
    errors.push(...validateSchema('api-schema', apiDoc, apiSchemaPath));
    errors.push(...validateSchema('mapping-logic', mappingDoc, mappingSchemaPath));
    if (errors.length === 0) {
      errors.push(...validate(uiDoc, apiDoc, mappingDoc));
    }
  } catch (err) {
    console.error(err.message);
    return 1;
  }

  if (errors.length > 0) {
    for (const error of errors) {
      console.error(`ERROR: ${error}`);
    }
    return 1;
  }

  console.log('OK: contracts are valid');
  return 0;
}

if (process.argv[1] === fileURLToPath(import.meta.url)) {
  process.exit(main());
}

export { validate, validateUniqueness, validateSchema };
